### ISSUE TITLE
Reduce need for query file parsing, to improve performance

### ISSUE THREAD

**Issue by Xophmeister** [tweag/topiary#443]
**Is your feature request related to a problem? Please describe.**
Topiary spends a lot of its time parsing the query file, which can lead to slow downs.

(e.g., See #442, raised by @nojb)

**Describe the solution you'd like**
There are a number of potential solutions to this. For example:

1. Cache or precompute the query files, so they don't need to be parsed on every invocation. (See also https://github.com/tree-sitter/tree-sitter/issues/1942)
2. Run Topiary in a "server mode", where query files are parsed on first access and are then retained in memory, for subsequent formatting.

**Comment by dumblob** [comment]
As a noob here, regarding the point (2) - is the query "cache" in memory then updated after some changes?

If so, is it done incrementally or "brute force (parse everything as in case of the first access)"?

**Comment by ErinvanderVeen** [comment]
Yes! This is exactly true! The `tree_sitter::Query` structure (or actually the one from `tree_sitter_facade`) is created once for the lifetime of the process. However, if we were to implement option 2, the server process would stay alive when a query is updated. Because, keep in mind that, with Topiary, you can define your own query files!

So in that case, the server process would definitely need to be informed of a Query file change.

`tree-sitter` currently does not allow incremental building of the Query, so it would have to be done from scratch.

Regardless, we are currently banking on option 2, which would also be beneficial for other users `tree_sitter`.
